### DISCUSSION TITLE
Shared References when Folding with a Row or Column?

### DISCUSSION THREAD

**Discussion by cryptoquick** [hecrj/iced#1382] [Q&A] [answered]
### Is there an existing issue for this?

- [X] I have searched the existing issues.

### Is this issue related to iced?

- [X] My hardware is compatible and my graphics drivers are up-to-date.

### What happened?

Hi, I'm trying to create a layout kind of like the Pinterest "masonry" layout. I have a 2D vec of image thumbnails I'm using fold to iterate over, and running into issues around shared references.

```rust
let row = row().width(Length::Fill).height(Length::Fill);
let row = image_grid.iter().fold(row, |r, images| {
    let column = column().spacing(row_spacing).width(col_width);
    let column = images.iter().fold(column, |c, path| {
        let el = self.publish_thumbs.get(path).as_ref().map(|thumb| {
            image(Handle::from_pixels(
                thumb.val().metadata.width_px,
                thumb.val().metadata.height_px,
                thumb.val().image.to_vec(),
            ))
        });

        if let Some(el) = el {
            c.push(el);
        }
        c
    });

    r.push(column);
    r
});

let scrollable = scrollable(row);
```

```
use of moved value: `c`
value used here after move

use of moved value: `r`
value used here after move
```

### What is the expected behavior?

I was hoping to not need to use smart pointers, but I may need to. It's just hard to move things out of the smart pointers after using them.

### Version

0.4

### Operative System

Linux

**Comment by cryptoquick** [comment]
It'd be really nice if there was an impl for `row(Vec<Element>)` and such, like this, as a substitute for the `with_children` methods:

```rust
let row = row(image_grid.iter().map(|images| {
    column(images.iter().filter_map(|path| {
        self.publish_thumbs.get(path).as_ref().map(|thumb| {
            image(Handle::from_pixels(
                thumb.val().metadata.width_px,
                thumb.val().metadata.height_px,
                thumb.val().image.to_vec(),
            ))
        })
    }))
    .spacing(row_spacing)
    .width(col_width)
}))
.width(Length::Fill)
.height(Length::Fill);
```

**Reply by hecrj** [reply]
You can still use `Row::with_children` by importing `iced::pure::widget::Row`.

**Comment by hecrj** [comment, marked as answer]
`push` takes `self` and returns `Self`, so you can just write:

```rust
// ...

       if let Some(el) = el {
            c.push(el)
        } else {
            c
        }
    });

    r.push(column)
});
```
